### PR TITLE
feat: initial focus disabling

### PR DESCRIPTION
**Pull request by Lightllord**
## PR Checklist

Please check if your PR fulfills the following requirements:

-   [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
-   [x] Tests for the changes have been added (for bug fixes / features)
-   [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

initial focus directive can't be disabled with input parameter

Issue Number: N/A

## What is the new behavior?

initial focus directive can be disabled with input parameter

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/cdk/directives/src/initial-focus/initial-focus.directive.ts [hunk 1]
```diff
@@ -16,15 +19,23 @@ export class InitialFocusDirective implements AfterViewInit, OnDestroy {
     }
 
     public ngAfterViewInit(): void {
-        this.ngZone.runOutsideAngular((): void => {
-            window.clearInterval(this.timeoutId!);
-            // eslint-disable-next-line no-restricted-properties
-            this.timeoutId = window.setInterval((): void => this.focus(), this.focusDelay);
-        });
+        this.decideAndTryToFocus();
```
Comment (splincode) [inline packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:22]: move to `@angular-ru/cdk/rxjs`;

```ts
export function delayedOnce(time: number) {
  return timer(time).pipe(take(1));
}
```

```ts
export function outsideZone<T>(zone: NgZone) {
  return function(source: Observable<T>) {
    return new Observable(observer => {
      let sub: Subscription | null = null;
      
      zone.runOutsideAngular(() => {
        sub = source.subscribe(observer);
      });

      return sub;
    });
  };
}
```

and then

```ts
public ngAfterViewInit(): void {
   if (this.isCanBeFocused) {
      delayedOnce(this.delay).pipe(
          outsideZone(this.ngZone),
          takeUntil(this.unsubscribe$)
       ).subscribe(() => this.focus());
   }
}
```

**Review by splincode — CHANGES_REQUESTED** [review]

File: packages/cdk/directives/src/initial-focus/initial-focus.directive.ts (hunk 1 above)
Comment (splincode) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:22]: `rxjs timer will by itself be completed right after the first value was emitted`

Of course, however, there are a lot of such code sections in the project. We could be reused `delayedOnce` if we need in any place.

`I'll just get rid of take(1)` Could you explain in more detail?

Comment (Lightllord) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:22]: https://stackblitz.com/edit/typescript-11wmqe?file=index.ts

Comment (Lightllord) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:22]: about outsideZone
pls, look at difference between two outsideZone pipes here:

(нажать кнопку, смотреть в консоль)
https://stackblitz.com/edit/angular-zone-rxjs-patch-test-tx3ff9?file=src/app/app.component.ts
надо оставить второй вариант, ну и придумать - как бы пайп протестировать нормально

Comment (splincode) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:22]: > https://stackblitz.com/edit/typescript-11wmqe?file=index.ts

 oh yes, it looks like I got it working yesterday and confused it with the interval operator

Comment (splincode) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:22]: > надо оставить второй вариант, ну и придумать - как бы пайп протестировать нормально

could you make it a separate PR?

Comment (Lightllord) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:22]: yes, I gonna leave this PR as it is and pull outsideZone operator into a separate one
so i appreciate your accepts)

File: packages/cdk/directives/src/initial-focus/initial-focus.directive.ts [hunk 2]
```diff
@@ -16,15 +19,23 @@ export class InitialFocusDirective implements AfterViewInit, OnDestroy {
...
     }
 
     public ngOnDestroy(): void {
-        window.clearInterval(this.timeoutId!);
+        this.unsubscribe$.next(true);
+    }
+
+    private decideAndTryToFocus(): void {
+        if (!this.disabled && !this.isFocused()) {
+            this.ngZone.runOutsideAngular((): void => {
```
Comment (splincode) [inline packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:31]: I'll be waiting before you add outsideZone operator

**Review by splincode — CHANGES_REQUESTED** [review]

File: packages/cdk/directives/src/initial-focus/initial-focus.directive.ts (hunk 2 above)
Comment (Lightllord) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:31]: I'd better have it fixed at a refactoring task
initial-focus runs smoothly 
need it for another task

Comment (splincode) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:31]: Ok https://github.com/Angular-RU/angular-ru-sdk/issues/1148

**Review by splincode — APPROVED** [review]

File: packages/cdk/directives/src/initial-focus/initial-focus.directive.ts [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
 
 @Directive({ selector: '[initialFocus]' })
 export class InitialFocusDirective implements AfterViewInit, OnDestroy {
     private readonly className: string = 'initial-focused';
-    private timeoutId: Nullable<number> = null;
-    @Input() public focusDelay: number = MIN_DELAY;
+    private readonly unsubscribe$: Subject<boolean> = new Subject<boolean>();
+    @Input('focusDelay') public delay: number = MIN_DELAY;
+    @Input('focusDisabled') public disabled: boolean = false;
+    @Input('type') public type?: string;
```
Comment (splincode) [inline packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:13]: `@Input() public type?: string;`

Comment (splincode) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:13]: @Lightllord check it

Comment (Lightllord) [reply packages/cdk/directives/src/initial-focus/initial-focus.directive.ts:13]: did it like that just to not change variables in the logic
fixed

File: packages/cdk/docs/directives.md [hunk 4]
```diff
@@ -1,6 +1,11 @@
 #### `@angular-ru/cdk/directives`
 
 -   `InitialFocusDirective, InitialFocusModule`
+-   Selector: `initialFocus`
+
+API:
+
+-   `Input('focusDisabled) disabled: boolean` - default: `false`
```
Comment (splincode) [inline packages/cdk/docs/directives.md:8]: rename please

**Review by splincode — APPROVED** [review]

**Comment by Lightllord** [comment]
Не понял че сломалось

**Comment by splincode** [comment]
@Lightllord fixed

**Pull request merged**